clientv3/retry.go

```diff
@@ -16,72 +16,256 @@ package clientv3
+	// converts to grpc/status.(*statusError) by grpc/toRPCErr
+	// (e.g. transport.ErrConnClosing when server closed transport, failing node)
+	// if known status error from gRPC with following codes,
+	// then endpoint switch and retry
+	if s.Code() == codes.Unavailable ||
+		s.Code() == codes.Internal ||
+		s.Code() == codes.DeadlineExceeded {
+		unhealthy, connectWait, switchEp, retryEp = true, false, true, true
+		switch s.Message() {
+		case "there is no address available":
+			// pinned == "" or endpoint unpinned right before/during RPC send
+			// no need to mark empty endpoint
+			unhealthy = false
+
+			// RPC was sent with empty pinned address:
+			//   1. initial connection has not been established (never pinned)
+			//   2. an endpoint has just been unpinned from an error
+			// Both cases expect to pin a new endpoint when Up is called.
+			// Thus, should be retried.
+			retryEp = true
+			if logger.V(4) {
+				logger.Infof("clientv3/do: there was no pinned endpoint (will be retried)")
+			}
+
+			// case A.
+			// gRPC is being too slow to start transport (e.g. errors, backoff),
+			// then endpoint switch. Otherwise, it can block forever to connect wait.
+			// This case is safe to reset/drain all current connections.
+			switchEp = true
+
+			// case B.
+			// gRPC is just about to establish transport within a moment,
+			// where endpoint switch would enforce connection drain on the healthy
+			// endpoint, which otherwise could be pinned and connected.
+			// The healthy endpoint gets pinned but unpinned right after with
+			// an error "grpc: the connection is drained", from endpoint switch.
+			// Thus, not safe to trigger endpoint switch; new healthy endpoint
+			// will be marked as unhealthy and not be pinned for another >3 seconds.
+			// Not acceptable when this healthy endpoint was the only available one!
+			// Workaround is wait up to dial timeout, and then endpoint switch
+			// only when the connection still has not been "Up" (still no pinned endpoint).
+			// TODO: remove this when gRPC has better way to track connection status
+			connectWait = true
+
+		case "grpc: the connection is closing",
+			"grpc: the connection is unavailable":
+			// gRPC v1.7 retries on these errors with FailFast=false
+			// etcd client sets FailFast=true, thus retry manually
+			unhealthy, connectWait, switchEp, retryEp = true, false, true, true
+
+		default:
+			// only retry immutable RPCs ("put at-most-once semantics")
+			// mutable RPCs should not be retried if unknown errors
+			if write {
+				unhealthy, connectWait, switchEp, retryEp = true, false, true, false
+			}
+		}
+	}
+
+	return unhealthy, connectWait, switchEp, retryEp, err
 }
 
-func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRPCFunc {
+type retryRPCFunc func(context.Context, rpcFunc) error
+
+const minDialDuration = 3 * time.Second
+
+func (c *Client) newRetryWrapper(write bool) retryRPCFunc {
```

Comment:
this becomes far more complicated than what we discussed. 

retry should only care about retry. we should put the initial wait connection logic in another func.